File: src/IsbInterop/IsbObjectResolver.vb
Imports System.Reflection
Imports Autofac
Imports Autofac.Core.Registration

''' <summary>
''' Резолвер объектов IS-Builder.
''' </summary>
Public Module IsbObjectResolver
  Private ReadOnly container As IContainer

  ''' <summary>
  ''' Получает экземпляр объекта.
  ''' </summary>
  ''' <typeparam name="TI">Интерфейс объекта.</typeparam>
  ''' <param name="rcwObject">COM-объект.</param>
  ''' <param name="scope">Область видимости.</param>
  ''' <returns>Объект.</returns>
  Public Function Resolve(Of TI As IBaseIsbObject)(rcwObject As Object, scope As IScope) As TI
    Dim result As TI
    Try
      result = container.Resolve(Of TI)(
        New TypedParameter(GetType(Object), rcwObject),
        New TypedParameter(GetType(IScope), scope))
    Catch generatedExceptionName As ComponentNotRegisteredException
      Dim [error] = String.Format(My.Resources.Resources.CannotCastResultToSpecifiedType, GetType(TI).Name)
      Throw New IsbInteropException([error])
    End Try

    Return result
  End Function

  ''' <summary>
  ''' Конструктор.
  ''' </summary>
  Sub New()
    Dim builder = New ContainerBuilder()

    builder.RegisterType(Of BaseIsbObjectImp)().[As](Of IBaseIsbObject)()

    builder.RegisterAssemblyTypes(Assembly.GetExecutingAssembly()).
      AssignableTo(GetType(BaseIsbObject)).
      [As](Function(t) t.GetInterfaces().
      Where(Function(i) i.Name.EndsWith(t.Name))).
      InstancePerDependency()

    container = builder.Build()
  End Sub

End Module

File: src/IsbInterop/IsbObjectResolver.cs
﻿using Autofac;
using Autofac.Core.Registration;
using IsbInterop.Properties;
using System.Linq;
using System.Reflection;

namespace IsbInterop
{
  /// <summary>
  /// Резолвер объектов IS-Builder.
  /// </summary>
  public static class IsbObjectResolver
  {
    private static readonly IContainer Container;

    /// <summary>
    /// Конструктор.
    /// </summary>
    static IsbObjectResolver()
    {
      var builder = new ContainerBuilder();

      builder.RegisterType<BaseIsbObjectImp>()
        .As<IBaseIsbObject>();

      builder.RegisterAssemblyTypes(Assembly.GetExecutingAssembly())
        .AssignableTo(typeof(BaseIsbObject))
        .As(t => t.GetInterfaces()
          .Where(i => i.Name.EndsWith(t.Name)))
        .InstancePerDependency();

      Container = builder.Build();
    }

    /// <summary>
    /// Получает экземпляр объекта.
    /// </summary>
    /// <typeparam name="TI">Интерфейс объекта.</typeparam>
    /// <param name="rcwObject">COM-объект.</param>
    /// <param name="scope">Область видимости.</param>
    /// <returns>Объект.</returns>
    //[System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.NoInlining)]
    public static TI Resolve<TI>(object rcwObject, IScope scope) where TI : IBaseIsbObject
    {
      TI result;
      try
      {
        result = Container.Resolve<TI>(
          new TypedParameter(typeof(object), rcwObject),
          new TypedParameter(typeof(IScope), scope));
      }
      catch (ComponentNotRegisteredException)
      {
        var error = string.Format(Resources.CannotCastResultToSpecifiedType, typeof(TI).Name);
        throw new IsbInteropException(error);
      }

      return result;
    }
  }
}
